src/Pact/Eval.hs

```diff
@@ -222,35 +242,156 @@ loadModule m bod1 mi g0 = do
+  defs <- unifiedDefs
   evaluatedDefs <- traverse evalConstRef defs
-  installModule (m,evaluatedDefs)
-  (evalRefs.rsNew) %= ((_mName m,(m,evaluatedDefs)):)
-  return (g1,modDefs1)
-
+  solvedDefs <- evaluateConstraints m evaluatedDefs mi
+  let md = ModuleData m solvedDefs
+  installModule md 
+  (evalRefs . rsNewModules) %= HM.insert _mName md
+  return (g1, modDefs1)
+loadModule i@Interface{..} body info gas0 = do
+  (gas1,idefs) <- case instantiate' body of
+    (TList bd _ _bi) -> do
+      let doDef (g,rs) t = do
+            dnm <- case t of
+              TDef {..} -> return $ Just _tDefName
+              TConst {..} -> return $ Just $ _aName _tConstArg
+              TSchema {..} -> return $ Just $ asString _tSchemaName
+              TUse {..} -> evalUse _tModuleName _tModuleHash _tInfo >> return Nothing
+              _ -> evalError (_tInfo t) "Invalid interface member"
+            case dnm of
+              Nothing -> return (g, rs)
+              Just dn -> do
+                g' <- computeGas (Left (_tInfo t,dn)) (GModuleMember i)
+                return (g + g',(dn,t):rs)
+      second HM.fromList <$> foldM doDef (gas0,[]) bd
+    t -> evalError (_tInfo t) "Malformed interface"
+  cs <- fmap stronglyConnCompR $ forM (HM.toList idefs) $ \(dn,d) ->
+      do
+        d' <- forM d $ \(f :: Name) -> do
+                dm <- resolveRef f
+                case (dm,f) of
+                  (Just t,_) -> return (Right t)
+                  (Nothing,Name fn _) ->
+                      case HM.lookup fn idefs of
+                        Just _ -> return (Left fn)
+                        Nothing -> evalError (_nInfo f) ("Cannot resolve \"" ++ show f ++ "\"")
+                  (Nothing,_) -> evalError (_nInfo f) ("Cannot resolve \"" ++ show f ++ "\"")
+        return (d', dn, mapMaybe (either Just (const Nothing)) $ toList d')
+  sorted <- forM cs $ \c -> case c of
+              AcyclicSCC v -> return v
+              --CyclicSCC [v] -> return v
+              CyclicSCC vs -> do
+                evalError (if null vs then info else _tInfo $ view _1 $ head vs) $ "Recursion detected: " ++ show vs
+  let unifiedDefs :: Eval e (HM.HashMap Text Ref)
+      unifiedDefs = foldl dresolve (pure HM.empty) sorted
+      -- insert a fresh Ref into the map, fmapping the Either to a Ref via 'unify'
+      dresolve :: Eval e (HM.HashMap Text Ref) -> (Term (Either Text Ref), Text, [Text]) -> Eval e (HM.HashMap Text Ref)
+      dresolve eds (d,dn,_) = do
+        ds <- eds
+        ud <- traverse (unify info ds) d 
+        return $ HM.insert dn (Ref ud) ds
+      evalConstRef = runPure . evalConsts
+  defs <- unifiedDefs
+  evaluatedDefs <- traverse evalConstRef defs
+  let md = ModuleData i evaluatedDefs
+  installModule md
+  (evalRefs . rsNewModules) %= HM.insert _interfaceName md
+  return (gas1, idefs)
 
+    
 
+-- | Evaluate interface constraints in module 
+--
+-- For each implemented member, there may be overlapping meta
```

Comment:
Verbose comment and weirdly over-focused on Meta -- there's a lot more work going on here. In case you can't tell I'm not into chatty comments.

Reply:
I've pared them down to the first line (I'm down with chatty comments 😄 )